docs/adrs/006-web-apis.md

```diff
@@ -0,0 +1,171 @@
+
+export const client = createPenumbraClient();
+```
+
+The flow of work with the `client` would be as follows:
+
+- User creates an instance of the `client`
+- They use the `client` to establish the injected connection, and the `client` saves the connection in its state
+- To fetch information from the wallet, the application passes the `client` instance to the fetch methods
+
+The idea of a single shared `client` is inspired by these popular TypeScript libraries:
+
+- [Prisma ORM](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/instantiate-prisma-client)
+- [Supabase](https://supabase.com/docs/reference/javascript/initializing)
+- [Viem](https://viem.sh/docs/clients/intro)
+
+## Injected connection
+
+The `client` must be able to connect to injected wallets. To make the API possible, the interface injected into the `window` object must be standardized:
+
+```ts
+export interface PenumbraProvider extends Readonly<PenumbraStateEventTarget> {
+  /** Should contain a URI at the provider's origin,
+   * serving a manifest describing this provider */
+  readonly manifest: string;
+
+  /** Call to acquire a `MessagePort` to this provider, asks user for approval if wasn't connected before */
+  readonly connect: () => Promise<MessagePort>;
+
+  /** Call to indicate the provider should discard approval of this origin. */
+  readonly disconnect: () => Promise<void>;
+
+  /** Should synchronously return the present connection state.
+   * - `true` indicates active connection.
+   * - `false` indicates connection is inactive.
+   */
+  readonly isConnected: () => boolean;
+
+  /** Synchronously return present injection state */
+  readonly state: () => PenumbraState;
+
+  /** Standard EventTarget methods emitting PenumbraStateEvent upon state changes */
+  readonly addEventListener: PenumbraEventTarget['addEventListener'];
+  /** Unsubscribes from the state change events. Provide the same callback as in `addEventListener` */
+  readonly removeEventListener: PenumbraEventTarget['removeEvenListener'];
+}
+```
+
+## Client API
+
+A globally shared `client` instance should be able to establish the connection with
+injected wallet providers and use the connection to query the blockchain. The API in this case would be:
+
+```ts
+import type { PenumbraService } from '@penumbra-zone/protobuf';
+import type { PromiseClient } from '@connectrpc/connect';
+
+interface PenumbraClient {
+  /**
+   * Connects to the injected provider, asks for user approval if wasn't connected before.
+   * If `providerUrl` argument is not provided, tries to connect to the first injected provider.
+   * Returns the manifest URL of the connected provider or throws an error otherwise.
+   */
+  readonly connect: (providerUrl?: string) => Promise<string>;
+
+  /**
+   * Tries to reconnect to the injected provider (the first one with `connected` state)
+   * without asking for user approval
+   */
+  readonly reconnect: (providerUrl?: string) => Promise<string>;
```

Comment:
i don't think this is appropriate.

1. any provider may identify itself as connected
2. multiple providers may identify themself as connected
3. a user likely wishes to use a specific provider

implementing any automatic selection behavior in this client api is likely to guide developers to create a negative user experience for a user that wishes to select a provider. so the api should avoid any automatic selection behavior.

requiring the developer to deliberately select a frontend will force developers to anticipate this ambiguity. at that point they can either

- prompt users to select a specific provider
- implement some technique for their dapp to remember
- encode their own priority

you've removed the `PenumbraClient.providers` method, but with that method a developer could easily implement a similar behavior by filtering the returned record by testing `isConnected` on each provider.

Reply:
Fair enough. Somehow I thought checking for manifest validity would be enough but I see know it isn't. Especially at the 'multiple providers may identify themself as connected' point.

Probably will delete it for now